In [1]:
from autograd import Node, Exp, Op

# 01. Basic Test

In [2]:
a = Node(3)
b = Node(2)
f = (2 * a + b**a) * b

In [3]:
f.backward()

In [4]:
a.grad, b.grad

(15.090354888959125, 38)

$a=3, b=2$

$f = (2a + b^a)b$

$\frac{\partial f}{\partial a} = 2b + b^{a+1}lnb = 4 + 16ln2 = 15.090$

$\frac{\partial f}{\partial b} = 2a + (a+1)b^a = 6 + 32 = 38$

# 02. Sigmoid (Custom)

In [5]:
def Sigmoid(value):
    return 1 / (1 + Exp(-value))

In [6]:
x = Node(1)
sig = Sigmoid(x)
sig.backward()

In [7]:
x.grad

0.19661193324148188

In [8]:
from math import exp
test = 1 / (1 + exp(-1))
print(test * (1 - test))

0.19661193324148185


# 03. Custom Operator Class

In [9]:
from math import sin, cos, pi

In [10]:
class Sin(Op):
    def __init__(self, left, right=None):
        self.op = "Sin"
        super().__init__(left, right)

    def forward(self):
        return sin(self.left.value)

    def backward(self, backgrad=None):
        if backgrad is None:
            backgrad = 1
        self.left.grad += backgrad * cos(self.left.value)
        if isinstance(self.left, Op):
            self.left.backward(self.left.grad)

In [11]:
x = Node(pi / 2)
y = Sin(2 * x)

$x = \pi / 2$

$y = sin(2x)$

$\frac{\partial y}{\partial x} = 2cos(2x) = -2$

In [12]:
y.backward()

In [13]:
x.grad

-2.0

# Other

In [14]:
x = Node(1)
f = Exp(-x**2)

$f = e^{-x^2}$

$\frac{\partial f}{\partial x} = -2xe^{-2x} = -\frac{2}{e} = -0.73576$

In [15]:
f.backward()

In [16]:
x.grad

-0.7357588823428847

In [17]:
-2 * 1 * exp(-1)

-0.7357588823428847